service/src/main/java/bio/terra/cbas/controllers/RunSetsApiController.java

```diff
@@ -82,107 +126,147 @@ public ResponseEntity<RunSetStateResponse> postRunSet(RunSetRequest request) {
-              errorMessage));
-      // TODO: After WM-1448 completes, add nicer-looking return response that returns
-      //  the error to the user immediately
-      return new ResponseEntity<>(HttpStatus.valueOf(e.getCode()));
-    } catch (JsonProcessingException e) {
-      // Should be super rare that jackson cannot convert an object to Json...
-      log.warn("Failed to convert inputs object to JSON", e);
-      return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
+  public static Optional<ResponseEntity<RunSetStateResponse>> validateRequest(
+      RunSetRequest request, int maxRecordIds) {
+    String errorMsg = "";
+
+    // check number of Record IDs in request is within allowed limit
+    int recordIdsSize = request.getWdsRecords().getRecordIds().size();
+    if (recordIdsSize > maxRecordIds) {
+      errorMsg =
+          "%s Record IDs submitted exceeds the maximum value of %s. "
+              .formatted(recordIdsSize, maxRecordIds);
+    }
+
+    // check that there are no duplicated Record IDs present in the request
+    List<String> recordIds = request.getWdsRecords().getRecordIds();
+    List<String> duplicateRecordIds =
+        recordIds.stream().filter(e -> Collections.frequency(recordIds, e) > 1).distinct().toList();
+    if (duplicateRecordIds.size() > 0) {
+      errorMsg +=
+          String.format("Duplicate Record ID(s) %s present in request.", duplicateRecordIds);
+    }
+
+    if (!errorMsg.isEmpty()) {
+      String finalErrorMsg = "Bad user request. Error(s): " + errorMsg;
+      log.warn(finalErrorMsg);
+      return Optional.of(
+          new ResponseEntity<>(
+              new RunSetStateResponse().errors(finalErrorMsg), HttpStatus.BAD_REQUEST));
     }
 
-    // Store the run:
+    return Optional.empty();
+  }
+
+  private RunStateResponse storeRun(
+      UUID runId,
+      String externalId,
+      RunSet runSet,
+      String recordId,
+      CbasRunStatus runState,
+      String errors) {
+    String additionalErrorMsg = "";
     int created =
         runDao.createRun(
             new Run(
                 runId,
-                workflowResponse.getRunId(),
+                externalId,
                 runSet,
-                dataTableRowId,
+                recordId,
                 OffsetDateTime.now(),
-                UNKNOWN,
+                runState,
                 OffsetDateTime.now(),
                 OffsetDateTime.now(),
-                null));
+                errors));
 
     if (created != 1) {
-      log.error(
-          "New workflow for record {} submitted to engine as {} but CBAS database INSERT returned '{} rows created'",
-          dataTableRowId,
-          workflowResponse.getRunId(),
-          created);
+      additionalErrorMsg =
+          String.format(
+              "CBAS failed to create new row for Record ID %s in %s state in database. INSERT returned '%s rows created'",
+              recordId, runState, created);
+      log.error(additionalErrorMsg);
```

Comment:
We probably want to escalate this error condition instead of logging and forgetting?

Reply:
Yes. Hence I am storing it in [this variable here](https://github.com/DataBiosphere/cbas/blob/1a4fc18d77fc47a34cf5b66f6993f5c90e09bc18/service/src/main/java/bio/terra/cbas/controllers/RunSetsApiController.java#L219) and adding it to the [Run's error field here](https://github.com/DataBiosphere/cbas/blob/1a4fc18d77fc47a34cf5b66f6993f5c90e09bc18/service/src/main/java/bio/terra/cbas/controllers/RunSetsApiController.java#L229).